README.md
Description
===========

Installs and configures MySQL client or server.

Requirements
============

Chef 0.10.10+.

Platform
--------

* Debian, Ubuntu
* CentOS, Red Hat, Fedora
* Mac OS X (Using homebrew)
* SmartOS (Using pkgin)

Tested on:

* Debian 5.0, 6.0
* Ubuntu 10.04-12.04
* CentOS 5.5-5.8, 6.2-6.3
* Mac OS X 10.7.2
* SmartOS

Cookbooks
---------

Requires Opscode's openssl cookbook for secure password generation.
See _Attributes_ and _Usage_ for more information.

The RubyGem installation in the `mysql::ruby` recipe requires a C
compiler and Ruby development headers to be installed in order to
build the mysql gem.

Requires `homebrew`
[cookbook](http://community.opscode.com/cookbooks/homebrew) on Mac OS
X.

Resources and Providers
=======================

The LWRP that used to ship as part of this cookbook has been
refactored into the
[database](http://community.opscode.com/cookbooks/database)
cookbook. Please see the README for details on updated usage.

Attributes
==========

See the `attributes/server.rb` or `attributes/client.rb` for default
values. Several attributes have values that vary based on the node's
platform and version.

* `node['mysql']['client']['packages']` - An array of package names
  that should be installed on "client" systems. This can be modified,
  e.g., to specify packages for Percona.
* `node['mysql']['server']['packages']` - An array of package names
  that should be installed on "server" systems. This can be modified,
  e.g., to specify packages for Percona.

* `node['mysql']['auto-increment-increment']` -
  auto-increment-increment value in my.cnf
* `node['mysql']['auto-increment-offset]` - auto-increment-offset
  value in my.cnf
* `node['mysql']['basedir']` - Base directory where MySQL is installed
* `node['mysql']['bind_address']` - Listen address for MySQLd
* `node['mysql']['conf_dir']` - Location for mysql conf directory
* `node['mysql']['confd_dir']` - Location for mysql conf.d style
  include directory
* `node['mysql']['data_dir']` - Location for mysql data directory
* `node['mysql']['ec2_path']` - location of mysql data_dir on EC2
  nodes
* `node['mysql']['allow_remote_root']` - Whether access as root must
  be allowed to connections originating in hosts other than localhost. 
  Defaults to false.
* `node['mysql']['root_network_acl']` - A network or host address or a 
  set of the former from which the user will be able to log in if root 
  remote access is allowed, default is from anywhere.
* `node['mysql']['grants_path']` - Path where the grants.sql should be
  written
* `node['mysql']['mysqladmin_bin']` - Path to the mysqladmin binary
* `node['mysql']['old_passwords']` - Sets the `old_passwords` value in
  my.cnf.
* `node['mysql']['pid_file']` - Path to the mysqld.pid file
* `node['mysql']['port']` - Liten port for MySQLd
* `node['mysql']['reload_action']` - Action to take when mysql conf
  files are modified. Also allows "reload" and "none".
* `node['mysql']['root_group']` - The default group of the "root" user
* `node['mysql']['service_name']` - The name of the mysqld service
* `node['mysql']['socket']` - Path to the mysqld.sock file
* `node['mysql']['use_upstart']` - Whether to use upstart for the
  service provider

Performance and other "tunable" attributes are under the
`node['mysql']['tunable']` attribute, corresponding to the same-named
parameter in my.cnf, and the default values are used. See
`attributes/server.rb`.

The following attributes are randomly generated passwords handled in
the `mysql::server` recipe, using the OpenSSL cookbook's
`secure_password` helper method. These are set using the `set_unless`
node attribute method, which allows them to be easily overridden e.g.
in a role.

* `node['mysql']['server_root_password']` - Set the server's root
  password
* `node['mysql']['server_repl_password']` - Set the replication user
  'repl' password
* `node['mysql']['server_debian_password']` - Set the debian-sys-maint
  user password

## Windows Specific

The following attributes are specific to Windows platforms.

* `node['mysql']['client']['version']` - The version of MySQL
  connector to install.
* `node['mysql']['client']['arch']` - Force 32 bit to work with the
  mysql gem
* `node['mysql']['client']['package_file']` - The MSI file for the
  mysql connector.
* `node['mysql']['client']['url']` - URL to download the mysql
  connector.
* `node['mysql']['client']['packages']` - Similar to other platforms,
  this is the name of the client package.
* `node['mysql']['client']['basedir']` - Base installation location
* `node['mysql']['client']['lib_dir']` - Libraries under the base location
* `node['mysql']['client']['bin_dir']` - binary directory under base location
* `node['mysql']['client']['ruby_dir']` - location where the Ruby
  binaries will be

Usage
=====

On client nodes, use the client (or default) recipe:

    { "run_list": ["recipe[mysql::client]"] }

This will install the MySQL client libraries and development headers
on the system.

On nodes which may use the `database` cookbook's mysql resources, also
use the ruby recipe. This installs the mysql RubyGem in the Ruby
environment Chef is using via `chef_gem`.

    { "run_list": ["recipe[mysql::client]", "recipe[mysql::ruby]"] }

If you need to install the mysql Ruby library as a package for your
system, override the client packages attribute in your node or role.
For example, on an Ubuntu system:

    {
      "mysql": {
        "client": {
          "packages": ["mysql-client", "libmysqlclient-dev","ruby-mysql"]
        }
      }
    }

This creates a resource object for the package and does the
installation before other recipes are parsed. You'll need to have the
C compiler and such (ie, build-essential on Ubuntu) before running the
recipes, but we already do that when installing Chef :-).

On server nodes, use the server recipe:

    { "run_list": ["recipe[mysql::server]"] }

On Debian and Ubuntu, this will preseed the mysql-server package with
the randomly generated root password in the recipe file. On other
platforms, it simply installs the required packages. It will also
create an SQL file, `/etc/mysql/grants.sql`, that will be used to set up
grants for the root, repl and debian-sys-maint users.

The recipe will perform a `node.save` unless it is run under
`chef-solo` after the password attributes are used to ensure that in
the event of a failed run, the saved attributes would be used.

On EC2 nodes, use the `server_ec2` recipe and the mysql data dir will
be set up in the ephmeral storage.

    { "run_list": ["recipe[mysql::server_ec2]"] }

When the `ec2_path` doesn't exist we look for a mounted filesystem
(eg, EBS) and move the data_dir there.

The client recipe is already included by server and 'default' recipes.

For more infromation on the compile vs execution phase of a Chef run:

* http://wiki.opscode.com/display/chef/Anatomy+of+a+Chef+Run

Chef Solo Note
==============

These node attributes are stored on the Chef
server when using `chef-client`. Because `chef-solo` does not
connect to a server or save the node object at all, to have the same
passwords persist across `chef-solo` runs, you must specify them in
the `json_attribs` file used. For example:

    {
      "mysql": {
        "server_root_password": "iloverandompasswordsbutthiswilldo",
        "server_repl_password": "iloverandompasswordsbutthiswilldo",
        "server_debian_password": "iloverandompasswordsbutthiswilldo"
      },
      "run_list":["recipe[mysql::server]"]
    }

License and Author
==================

- Author:: Joshua Timberman (<joshua@opscode.com>)
- Author:: AJ Christensen (<aj@opscode.com>)
- Author:: Seth Chisamore (<schisamo@opscode.com>)
- Author:: Brian Bianco (<brian.bianco@gmail.com>)
- Author:: Jesse Howarth (<him@jessehowarth.com>)
- Author:: Andrew Crump (<andrew@kotirisoftware.com>)
- Author:: Javier Perez-Griffo (<javier@tapp.in>)

Copyright:: 2009-2012 Opscode, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## v1.3.1:

* [COOK-1904] SmartOS support for Mysql

## v1.3.0:

**Important note for this release**

This version no longer installs Ruby bindings in the client recipe by
default. Use the ruby recipe if you'd like the RubyGem. If you'd like
packages from your distribution, use them in your application's
specific cookbook/recipe, or modify the client packages attribute.
This resolves the following tickets:

* COOK-932
* COOK-1009
* COOK-1384

Additionally, this cookbook now has tests (COOK-1439) for use under
test-kitchen.

The following issues are also addressed in this release.

* [COOK-1443] - MySQL (>= 5.1.24) does not support `innodb_flush_method`
  = fdatasync
* [COOK-1175] - Add Mac OS X support
* [COOK-1289] - handle additional tunable attributes
* [COOK-1305] - add auto-increment-increment and auto-increment-offset
  attributes
* [COOK-1397] - make the port an attribute
* [COOK-1439] - Add MySQL cookbook tests for test-kitchen support
* [COOK-1236] - Move package names into attributes to allow percona to
  free-ride
* [COOK-934] - remove deprecated mysql/libraries/database.rb, use the
  database cookbook instead.
* [COOK-1475] - fix restart on config change

## v1.2.6:

* [COOK-1113] - Use an attribute to determine if upstart is used
* [COOK-1121] - Add support for Windows
* [COOK-1140] - Fix conf.d on Debian
* [COOK-1151] - Fix server_ec2 handling /var/lib/mysql bind mount
* [COOK-1321] - Document setting password attributes for solo

## v1.2.4

* [COOK-992] - fix FATAL nameerror
* [COOK-827] - `mysql:server_ec2` recipe can't mount `data_dir`
* [COOK-945] - FreeBSD support

## v1.2.2

* [COOK-826] mysql::server recipe doesn't quote password string
* [COOK-834] Add 'scientific' and 'amazon' platforms to mysql cookbook

## v1.2.1

* [COOK-644] Mysql client cookbook 'package missing' error message is confusing
* [COOK-645] RHEL6/CentOS6 - mysql cookbook contains 'skip-federated' directive which is unsupported on MySQL 5.1

## v1.2.0

* [COOK-684] remove mysql_database LWRP

## v1.0.8:

* [COOK-633] ensure "cloud" attribute is available

## v1.0.7:

* [COOK-614] expose all mysql tunable settings in config
* [COOK-617] bind to private IP if available

## v1.0.6:

* [COOK-605] install mysql-client package on ubuntu/debian

## v1.0.5:

* [COOK-465] allow optional remote root connections to mysql
* [COOK-455] improve platform version handling
* externalize conf_dir attribute for easier cross platform support
* change datadir attribute to data_dir for consistency

## v1.0.4:

* fix regressions on debian platform
* [COOK-578] wrap root password in quotes
* [COOK-562] expose all tunables in my.cnf


metadata.rb
name		  "mysql"
maintainer        "Opscode, Inc."
maintainer_email  "cookbooks@opscode.com"
license           "Apache 2.0"
description       "Installs and configures mysql for client or server"
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version           "1.3.1"
recipe            "mysql", "Includes the client recipe to configure a client"
recipe            "mysql::client", "Installs packages required for mysql clients using run_action magic"
recipe            "mysql::server", "Installs packages required for mysql servers w/o manual intervention"
recipe            "mysql::server_ec2", "Performs EC2-specific mountpoint manipulation"

%w{ debian ubuntu centos suse fedora redhat scientific amazon freebsd windows mac_os_x smartos}.each do |os|
  supports os
end

depends "openssl"
depends "build-essential"
suggests "homebrew"
suggests "windows"
depends "pkgin"

attribute "mysql/server_root_password",
  :display_name => "MySQL Server Root Password",
  :description => "Randomly generated password for the mysqld root user",
  :default => "randomly generated"

attribute "mysql/bind_address",
  :display_name => "MySQL Bind Address",
  :description => "Address that mysqld should listen on",
  :default => "ipaddress"

attribute "mysql/data_dir",
  :display_name => "MySQL Data Directory",
  :description => "Location of mysql databases",
  :default => "/var/lib/mysql"

attribute "mysql/conf_dir",
  :display_name => "MySQL Conf Directory",
  :description => "Location of mysql conf files",
  :default => "/etc/mysql"

attribute "mysql/ec2_path",
  :display_name => "MySQL EC2 Path",
  :description => "Location of mysql directory on EC2 instance EBS volumes",
  :default => "/mnt/mysql"
  
attribute "mysql/allow_remote_root",
  :display_name => 'Allow remote root connections',
  :description => 'Whether access as root must be allowed to connections originating in hosts other than localhost',
  :default => 'false'  

attribute "mysql/root_network_acl",
  :display_name => 'Root network ACL',
  :description => 'A network or host address specified in MySQL grant format or a set of the former, from which the user will be able to log in',
  :default => '%'

attribute "mysql/reload_action",
  :display_name => "MySQL conf file reload action",
  :description => "Action to take when mysql conf files are modified",
  :default => "reload"

attribute "mysql/tunable",
  :display_name => "MySQL Tunables",
  :description => "Hash of MySQL tunable attributes",
  :type => "hash"

attribute "mysql/tunable/key_buffer",
  :display_name => "MySQL Tuntable Key Buffer",
  :default => "250M"

attribute "mysql/tunable/max_connections",
  :display_name => "MySQL Tunable Max Connections",
  :default => "800"

attribute "mysql/tunable/wait_timeout",
  :display_name => "MySQL Tunable Wait Timeout",
  :default => "180"

attribute "mysql/tunable/net_read_timeout",
  :display_name => "MySQL Tunable Net Read Timeout",
  :default => "30"

attribute "mysql/tunable/net_write_timeout",
  :display_name => "MySQL Tunable Net Write Timeout",
  :default => "30"

attribute "mysql/tunable/back_log",
  :display_name => "MySQL Tunable Back Log",
  :default => "128"

attribute "mysql/tunable/table_cache",
  :display_name => "MySQL Tunable Table Cache for MySQL < 5.1.3",
  :default => "128"

attribute "mysql/tunable/table_open_cache",
  :display_name => "MySQL Tunable Table Cache for MySQL >= 5.1.3",
  :default => "128"

attribute "mysql/tunable/max_heap_table_size",
  :display_name => "MySQL Tunable Max Heap Table Size",
  :default => "32M"

attribute "mysql/tunable/expire_logs_days",
  :display_name => "MySQL Exipre Log Days",
  :default => "10"

attribute "mysql/tunable/max_binlog_size",
  :display_name => "MySQL Max Binlog Size",
  :default => "100M"

attribute "mysql/client",
  :display_name => "MySQL Connector/C Client",
  :description => "Hash of MySQL client attributes",
  :type => "hash"

attribute "mysql/client/version",
  :display_name => "MySQL Connector/C Version",
  :default => "6.0.2"

attribute "mysql/client/arch",
  :display_name => "MySQL Connector/C Architecture",
  :default => "win32"

attribute "mysql/client/package_file",
  :display_name => "MySQL Connector/C Package File Name",
  :default => "mysql-connector-c-6.0.2-win32.msi"

attribute "mysql/client/url",
  :display_name => "MySQL Connector/C Download URL",
  :default => "http://www.mysql.com/get/Downloads/Connector-C/mysql-connector-c-6.0.2-win32.msi/from/http://mysql.mirrors.pair.com/"

attribute "mysql/client/package_name",
  :display_name => "MySQL Connector/C Registry DisplayName",
  :default => "MySQL Connector C 6.0.2"

attribute "mysql/client/basedir",
  :display_name => "MySQL Connector/C Base Install Directory",
  :default => "C:\\Program Files (x86)\\MySQL\\Connector C 6.0.2"

attribute "mysql/client/lib_dir",
  :display_name => "MySQL Connector/C Library Directory (containing libmysql.dll)",
  :default => "C:\\Program Files (x86)\\MySQL\\Connector C 6.0.2\\lib\\opt"

attribute "mysql/client/bin_dir",
  :display_name => "MySQL Connector/C Executable Directory",
  :default => "C:\\Program Files (x86)\\MySQL\\Connector C 6.0.2\\bin"

attribute "mysql/client/ruby_dir",
  :display_name => "Ruby Executable Directory which should gain MySQL support",
  :default => "system ruby"


recipes/ruby.rb
#
# Cookbook Name:: mysql
# Recipe:: ruby
#
# Author:: Jesse Howarth (<him@jessehowarth.com>)
# Author:: Jamie Winsor (<jamie@vialstudios.com>)
#
# Copyright 2008-2012, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

execute "apt-get update" do
  ignore_failure true
  action :nothing
end.run_action(:run) if node['platform_family'] == "debian"

node.set['build_essential']['compiletime'] = true
include_recipe "build-essential"
include_recipe "mysql::client"

if platform? 'smartos'
	include_recipe 'pkgin'
	node['mysql']['client']['packages'].each do |mysql_pack|
		pkgin_package mysql_pack do
			action :install
		end
	end
else
	node['mysql']['client']['packages'].each do |mysql_pack|
		resources("package[#{mysql_pack}]").run_action(:install)
	end
end

chef_gem "mysql"


recipes/client.rb
#
# Cookbook Name:: mysql
# Recipe:: client
#
# Copyright 2008-2011, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Include Opscode helper in Recipe class to get access
# to debian_before_squeeze? and ubuntu_before_lucid?
::Chef::Recipe.send(:include, Opscode::Mysql::Helpers)

case node['platform']
when "windows"
  package_file = node['mysql']['client']['package_file']
  remote_file "#{Chef::Config[:file_cache_path]}/#{package_file}" do
    source node['mysql']['client']['url']
    not_if { File.exists? "#{Chef::Config[:file_cache_path]}/#{package_file}" }
  end

  windows_package node['mysql']['client']['packages'].first do
    source "#{Chef::Config[:file_cache_path]}/#{package_file}"
  end
  windows_path node['mysql']['client']['bin_dir'] do
    action :add
  end
  def package(*args, &blk)
    windows_package(*args, &blk)
  end
when "smartos"
  include_recipe 'pkgin'

  def package(*args, &blk)
    pkgin_package(*args, &blk)
  end
when "mac_os_x"
  include_recipe 'homebrew'
end

node['mysql']['client']['packages'].each do |mysql_pack|
  package mysql_pack do
    action :install
  end
end

if platform? 'windows'
  ruby_block "copy libmysql.dll into ruby path" do
    block do
      require 'fileutils'
      FileUtils.cp "#{node['mysql']['client']['lib_dir']}\\libmysql.dll", node['mysql']['client']['ruby_dir']
    end
    not_if { File.exist?("#{node['mysql']['client']['ruby_dir']}\\libmysql.dll") }
  end
end


attributes/server.rb
#
# Cookbook Name:: mysql
# Attributes:: server
#
# Copyright 2008-2009, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['mysql']['bind_address']               = attribute?('cloud') ? cloud['local_ipv4'] : ipaddress
default['mysql']['port']                       = 3306

case node["platform"]
when "centos", "redhat", "fedora", "suse", "scientific", "amazon"
  default['mysql']['server']['packages']      = %w{mysql-server}
  default['mysql']['service_name']            = "mysqld"
  default['mysql']['basedir']                 = "/usr"
  default['mysql']['data_dir']                = "/var/lib/mysql"
  default['mysql']['root_group']              = "root"
  default['mysql']['mysqladmin_bin']          = "/usr/bin/mysqladmin"
  default['mysql']['mysql_bin']               = "/usr/bin/mysql"

  set['mysql']['conf_dir']                    = '/etc'
  set['mysql']['confd_dir']                   = '/etc/mysql/conf.d'
  set['mysql']['socket']                      = "/var/lib/mysql/mysql.sock"
  set['mysql']['pid_file']                    = "/var/run/mysqld/mysqld.pid"
  set['mysql']['old_passwords']               = 1
  set['mysql']['grants_path']                 = "/etc/mysql_grants.sql"
  # RHEL/CentOS mysql package does not support this option.
  set['mysql']['tunable']['innodb_adaptive_flushing'] = false
when "freebsd"
  default['mysql']['server']['packages']      = %w{mysql55-server}
  default['mysql']['service_name']            = "mysql-server"
  default['mysql']['basedir']                 = "/usr/local"
  default['mysql']['data_dir']                = "/var/db/mysql"
  default['mysql']['root_group']              = "wheel"
  default['mysql']['mysqladmin_bin']          = "/usr/local/bin/mysqladmin"
  default['mysql']['mysql_bin']               = "/usr/local/bin/mysql"

  set['mysql']['conf_dir']                    = '/usr/local/etc'
  set['mysql']['confd_dir']                   = '/usr/local/etc/mysql/conf.d'
  set['mysql']['socket']                      = "/tmp/mysqld.sock"
  set['mysql']['pid_file']                    = "/var/run/mysqld/mysqld.pid"
  set['mysql']['old_passwords']               = 0
  set['mysql']['grants_path']                 = "/var/db/mysql/grants.sql"
when "smartos"  
  default['mysql']['server']['packages']      = %w{tcp_wrappers mysql-server}
  default['mysql']['package_name']            = "mysql-server"
  default['mysql']['service_name']            = "mysql"
  default['mysql']['basedir']                 = "/opt/local"
  default['mysql']['data_dir']                = "/var/mysql"
  default['mysql']['root_group']              = "root"
  default['mysql']['mysqladmin_bin']          = "/opt/local/bin/mysqladmin"
  default['mysql']['mysql_bin']               = "/opt/local/bin/mysql"
  default['mysql']['root_group']              = "mysql"

  set['mysql']['conf_dir']                    = '/etc'
  set['mysql']['confd_dir']                   = '/etc/mysql/conf.d'
  set['mysql']['socket']                      = "/tmp/mysql.sock"
  set['mysql']['pid_file']                    = "/var/mysql/mysql.pid"
  set['mysql']['old_passwords']               = 0
  set['mysql']['grants_path']                 = "/opt/local/share/mysql/grants.sql"
when "windows"
  default['mysql']['server']['packages']      = ["MySQL Server 5.5"]
  default['mysql']['version']                 = '5.5.21'
  default['mysql']['arch']                    = 'win32'
  default['mysql']['package_file']            = "mysql-#{mysql['version']}-#{mysql['arch']}.msi"
  default['mysql']['url']                     = "http://www.mysql.com/get/Downloads/MySQL-5.5/#{mysql['package_file']}/from/http://mysql.mirrors.pair.com/"

  default['mysql']['service_name']            = "mysql"
  default['mysql']['basedir']                 = "#{ENV['SYSTEMDRIVE']}\\Program Files (x86)\\MySQL\\#{mysql['server']['packages'].first}"
  default['mysql']['data_dir']                = "#{mysql['basedir']}\\Data"
  default['mysql']['bin_dir']                 = "#{mysql['basedir']}\\bin"
  default['mysql']['mysqladmin_bin']          = "#{mysql['bin_dir']}\\mysqladmin"
  default['mysql']['mysql_bin']               = "#{mysql['bin_dir']}\\mysql"

  default['mysql']['conf_dir']                = "#{mysql['basedir']}"
  default['mysql']['old_passwords']           = 0
  default['mysql']['grants_path']             = "#{mysql['conf_dir']}\\grants.sql"
when "mac_os_x"
  default['mysql']['server']['packages']      = %w{mysql}
  default['mysql']['basedir']                 = "/usr/local/Cellar"
  default['mysql']['data_dir']                = "/usr/local/var/mysql"
  default['mysql']['root_group']              = "admin"
  default['mysql']['mysqladmin_bin']          = "/usr/local/bin/mysqladmin"
  default['mysql']['mysql_bin']               = "/usr/local/bin/mysql"
else
  default['mysql']['server']['packages']      = %w{mysql-server}
  default['mysql']['service_name']            = "mysql"
  default['mysql']['basedir']                 = "/usr"
  default['mysql']['data_dir']                = "/var/lib/mysql"
  default['mysql']['root_group']              = "root"
  default['mysql']['mysqladmin_bin']          = "/usr/bin/mysqladmin"
  default['mysql']['mysql_bin']               = "/usr/bin/mysql"

  set['mysql']['conf_dir']                    = '/etc/mysql'
  set['mysql']['confd_dir']                   = '/etc/mysql/conf.d'
  set['mysql']['socket']                      = "/var/run/mysqld/mysqld.sock"
  set['mysql']['pid_file']                    = "/var/run/mysqld/mysqld.pid"
  set['mysql']['old_passwords']               = 0
  set['mysql']['grants_path']                 = "/etc/mysql/grants.sql"
end

if attribute?('ec2')
  default['mysql']['ec2_path']    = "/mnt/mysql"
  default['mysql']['ebs_vol_dev'] = "/dev/sdi"
  default['mysql']['ebs_vol_size'] = 50
end

default['mysql']['reload_action'] = "restart" # or "reload" or "none"

default['mysql']['use_upstart'] = node.platform?("ubuntu") && node.platform_version.to_f >= 10.04

default['mysql']['auto-increment-increment']        = 1
default['mysql']['auto-increment-offset']           = 1

default['mysql']['allow_remote_root']               = false
default['mysql']['root_network_acl']                = '%'
default['mysql']['tunable']['back_log']             = "128"
default['mysql']['tunable']['key_buffer']           = "256M"
default['mysql']['tunable']['max_allowed_packet']   = "16M"
default['mysql']['tunable']['max_connections']      = "800"
default['mysql']['tunable']['max_heap_table_size']  = "32M"
default['mysql']['tunable']['myisam_recover']       = "BACKUP"
default['mysql']['tunable']['net_read_timeout']     = "30"
default['mysql']['tunable']['net_write_timeout']    = "30"
default['mysql']['tunable']['table_cache']          = "128"
default['mysql']['tunable']['table_open_cache']     = "128"
default['mysql']['tunable']['thread_cache']         = "128"
default['mysql']['tunable']['thread_cache_size']    = 8
default['mysql']['tunable']['thread_concurrency']   = 10
default['mysql']['tunable']['thread_stack']         = "256K"
default['mysql']['tunable']['wait_timeout']         = "180"

default['mysql']['tunable']['log_bin']                         = nil
default['mysql']['tunable']['log_bin_trust_function_creators'] = false
default['mysql']['tunable']['relay_log']                       =  nil
default['mysql']['tunable']['log_slave_updates']               = false
default['mysql']['tunable']['sync_binlog']                     = 0
default['mysql']['tunable']['skip_slave_start']                = false

default['mysql']['tunable']['log_error']                       = nil
default['mysql']['tunable']['log_queries_not_using_index']     = true
default['mysql']['tunable']['log_bin_trust_function_creators'] = false

default['mysql']['tunable']['innodb_buffer_pool_size']         = "128M"
default['mysql']['tunable']['innodb_log_file_size']            = "5M"
default['mysql']['tunable']['innodb_additional_mem_pool_size'] = "8M"
default['mysql']['tunable']['innodb_data_file_path']           = "ibdata1:10M:autoextend"
default['mysql']['tunable']['innodb_flush_log_at_trx_commit']  = "1"
default['mysql']['tunable']['innodb_flush_method']             = false
default['mysql']['tunable']['innodb_log_buffer_size']          = "8M"

default['mysql']['tunable']['query_cache_limit']    = "1M"
default['mysql']['tunable']['query_cache_size']     = "16M"

default['mysql']['tunable']['log_slow_queries']     = "/var/log/mysql/slow.log"
default['mysql']['tunable']['long_query_time']      = 2

default['mysql']['tunable']['expire_logs_days']     = 10
default['mysql']['tunable']['max_binlog_size']      = "100M"
